### ISSUE TITLE
security: umbrella tracker for adversarial-audit findings

### ISSUE THREAD

**Issue by yingbull** [carlos-emr/carlos#1636]
## Summary

Umbrella tracking ticket for all security findings discovered via **multi-pass adversarial validation** of GitHub code-scanning alerts (April 2026). This is a meta-issue capturing findings that were either (a) missed by the scanner entirely, or (b) scanner-flagged but required cross-file / cross-layer tracing to confirm as real vulnerabilities rather than false positives.

**Methodology**: ~20 parallel agents performing 4 passes of source-to-sink tracing, adversarial input construction, and consumer-layer auditing against the original 167 scanner alerts, expanding to neighboring code via pattern-sibling discovery.

**Result**: 7 classifications flipped (FP → TP) across passes, plus ~15 additional vulnerabilities discovered outside the scanner's coverage. Of these, 10 are already tracked in #1634 and #1635. This ticket captures the remaining 8+ items and serves as the master inventory.

---

## Classification legend

- 🔴 **TP-flip** — scanner raised this but the original classification was wrong (either FP-flipped-to-TP or TP-flipped-to-FP)
- 🆕 **Adversarial discovery** — scanner did not flag this; found via pattern-sibling analysis or cross-file trace
- 🟡 **Severity adjustment** — already triaged but severity needs updating based on deeper analysis

---

### In issue #1634 — "Low-risk security fixes bundle"
- [x] 🔴 `navigation.jsp:405,414` — stored XSS via `bean.userName` in `onClick` JS context (Fix 1)
- [x] 🔴 `HashAudit.java:49` — MD5 for signed-note audit integrity (Fix 2)
- [x] 🔴 `ReportReassign2Action.java:256` — protocol-relative open redirect via `getRequestURI()` (Fix 3)
- [x] 🆕 `CaseManagementEntry2Action.isValidInternalRedirect` — accepts `/\evil.example` backslash bypass (Fix 5)
- [x] 🆕 `provider/obarchecklistedit_99_12.jsp` — unpatched duplicate of the hardened sibling; missing admin gate entirely (Fix 7)

### In issue #1635 — "JSP output-encoding hardening bundle"
- [x] 🆕 `casemgmt/unlockAjax.jsp` — 4 stored XSS sinks (note body, provider name, encounter type, issue description)
- [x] 🆕 `provider/clients.jsp:156,160` — stored XSS on OAuth client name / verifier
- [x] 🆕 `fieldnoteselect.jsp:108-146` — stored XSS on eform name/subject/filename
- [x] 🆕 `demographic/demographicAudit.jsp:114-118` — inconsistent encoding; audit-log fields emitted raw
- [x] 🆕 `displayHRMReport.jsp:522` — stored XSS via HRM content from external hospital systems

---

### Category A — Stored XSS in JSP fragments (multi-field, not single-line)

Each of these requires encoding multiple fields per file; they were excluded from #1635 because the diffs are larger and the file-level review benefit from dedicated scoping. The underlying fix pattern is identical (`Encode.forHtml()` / `Encode.forHtmlAttribute()` / `Encode.forUriComponent()`).

- [ ] **A1 — `src/main/webapp/encounter/Index2.jsp:175,213`** (caisi module)
  - `bean.userName` concatenated into URL fed to `response.sendRedirect` without URL-encoding
  - Fix: `URLEncoder.encode(bean.userName, StandardCharsets.UTF_8)`
  - Sibling of #1634 Fix 1; same `EctSessionBean` source

- [ ] **A2 — `src/main/webapp/tickler/ticklerDemoMain.jsp`**
  - Lines 900, 901, 903, 923, 928, 987-988, 1001: `d.getLastName()`, `d.getFirstName()`, `provider`, `taskAssignedTo`, `t.getMessage()` (user-entered tickler message), `tc.getProvider().getLastName()`, `tc.getProvider().getFirstName()`, `tc.getMessage()` — all raw
  - Fix: wrap each with `Encode.forHtml()`
  - Highest-impact items: tickler messages (user-entered text) — classic stored XSS vector

- [ ] **A3 — `src/main/webapp/casemgmt/ticklers.jsp`**
  - Lines 123, 128, 130, 132-133: `provider_name`, `assignee_name`, `status` raw emission
  - Line 130 area: `<c:out escapeXml="false" value="${tickler.message}"/>` — `escapeXml="false"` anti-pattern
  - Fix: `<c:out escapeXml="true" value="${tickler.message}"/>` (or `${e:forHtml(tickler.message)}` using Jakarta EL function)

- [ ] **A4 — `src/main/webapp/oscarRx/Preview2.jsp`**
  - Lines 435, 439-444, 673: `doctorName`, `provider.getClinicName/Address/City/Province/Postal/PractitionerNo`, `fullOutLine` (prescription text) — all raw
  - Line 494: `<c:out escapeXml="false" value="${infirmaryView_programAddress}"/>` — anti-pattern
  - Fix: wrap all with `Encode.forHtml()`; convert `escapeXml="false"` to `true` or pre-encode content

- [ ] **A5 — `src/main/webapp/lab/CA/BC/index.jsp:178,181,185,189`**
  - HL7-sourced data (`obr.getOrderingProvider()`, `getResultCopiesTo()`, `demo.getFullName()`, `demo.getDemographicNo()`) rendered without encoding
  - Inconsistent with sibling `lab/CA/BC/report.jsp` which uses `Encode.forHtml(…).replaceAll(...)`
  - Fix: apply the same pattern from `report.jsp`

- [ ] **A6 — `src/main/webapp/WEB-INF/jsp/caisicore/DefaultEncounterIssueList.jsp:127-130`**
  - Same `out.print("<td>" + provider.getLastName() + "</td>")` raw concatenation pattern as unlockAjax
  - Fix: wrap `provider.getLastName()`, `provider.getFirstName()`, `sb.toString()` (issue descriptions) with `Encode.forHtml()`

---

### Category B — Reflected XSS via `<c:url>` + `escapeXml="false"` anti-pattern

JSTL `<c:url>` does **not** URL-encode values interpolated directly into its `value` attribute (only values passed via nested `<c:param>`). When the resulting URL is emitted inside an `onclick="…('<c:out escapeXml=false value='${notesURL}'/>')"` — a JavaScript single-quoted string inside an HTML attribute — unencoded `'` or `)` in request parameters break out of the JS context.

- [ ] **B1 — `src/main/webapp/casemgmt/search.jsp:164, 173, 186, 199, 261, 269, 276, 282`**
- [ ] **B2 — `src/main/webapp/casemgmt/CaseManagementView.jsp:382, 394, 450-451, 464-465, 479-480, 555-556, 563-564, 572-573, 624, 636, 650`**

**Attack**: `?demographicNo=');alert(1)//` reflected into `onclick="popupNotePage('…&demographicNo=');alert(1)//…')"` → script executes in the referrer's session.

**Fix strategy** (two options):
1. **Local** — use `Encode.forJavaScript()` on the full URL before emission into the `onclick` string:
   ```jsp
   <c:set var="notesURL"><c:url value="…${param.demographicNo}…"/></c:set>
   onclick="popupNotePage('<%= Encode.forJavaScript((String)pageContext.getAttribute("notesURL")) %>')"
   ```
2. **Better** — convert to event-delegation JavaScript: emit only the data as a safe attribute (`data-demographic-no="<%=Encode.forHtmlAttribute(...)%>"`) and attach the click handler via JS, reading the dataset. Removes the JS-in-HTML-attribute pattern entirely.

**Scope**: This is a JSTL anti-pattern that appears in more than these 2 files; a project-wide grep for `escapeXml="false"` on EL-interpolated URLs is warranted. Consider adding a CI lint rule.

---

### Category C — Crypto / key management

- [ ] **C1 — `EncryptionUtils.java:320-328`** (🆕 adversarial)
  - AES-GCM key loaded from `CarlosProperties.encryption.util.secret.key` via simple base64 decode
  - No KDF (if operator enters a low-entropy base64 string, key has low entropy)
  - No salt
  - No key rotation mechanism
  - Single static key encrypts all GCM-protected data
  - Risk: filesystem disclosure of `carlos.properties` = total compromise of all encrypted PHI at rest
  - **Fix plan** (needs design discussion):
    - [ ] Move key storage to a secret manager (HashiCorp Vault, AWS KMS, Azure Key Vault, or at minimum an OS-level keystore)
    - [ ] Add a KDF layer (HKDF or PBKDF2) so operator-supplied passphrases are strengthened
    - [ ] Define a rotation strategy: key versioning in ciphertext format, periodic re-encryption batch job
    - [ ] Update `EncryptionUtils` to accept a key provider interface rather than reading the property directly
  - **Not a quick fix**. Requires architectural decision and possibly a new dependency.

- [ ] **C2 — `LabUpload2Action.java:217`** (🟡 severity adjustment) — tracked in original scanner alert `java/rsa-without-oaep` (alert #148)
  - Bleichenbacher oracle analysis: error paths are partially unified (single `catch(Exception){return null;}`), reducing response-code differential
  - Timing differential likely remains observable
  - Exploitation requires thousands of oracle queries, feasible only from lab-partner IP or open endpoint
  - **Severity**: Medium → **Low** (given unified error handling + bounded attack surface)
  - **Long-term fix**: coordinate with upstream lab senders to migrate to RSA-OAEP; add rate limiting + constant-time error paths as interim mitigation
  - **Correct the misleading in-code comment** (currently claims "decrypt-only limits attack surface" — this is technically incorrect; Bleichenbacher IS a decrypt-only oracle)

- [ ] **C3 — `HashAudit.java`** (🟡 severity adjustment) — tracked via #1634 Fix 2
  - Post-adversarial verification: the `hash_audit` table is **write-only** in production code (no `verify()` or read path exists today)
  - Severity: Medium → **Low / Informational** (algorithm is wrong for intended use but not exploitable as a bypass in current code)
  - #1634 Fix 2 still applies; this is a severity note, not a scope change

---

### Category D — Authorization / hardening gaps

- [ ] **D1 — `ws/OscarJobService.java::saveJobType`** (🆕 adversarial)
  - REST endpoint for updating `OscarJobType` records lacks explicit `SecurityInfoManager.hasPrivilege("_admin", "w", null)` check
  - Currently protected only by OAuth authentication (not by RBAC)
  - Impact: any OAuth-authenticated client can set `OscarJobType.className`, triggering `Class.forName(className).newInstance()` downstream in `OscarJobUtils.java:122` — potential RCE via static initializer or constructor side effects
  - Defense-in-depth gap; UI-level `_admin` gating does not protect the REST surface
  - **Fix**: add `hasPrivilege("_admin","w")` guard at the top of `saveJobType`

- [ ] **D2 — `AddEditDocument2Action.java:270`** (🟡 severity caveat)
  - Previously classified FP for open-redirect (host is hardcoded), but params `function`, `functionid`, `curUser`, `appointmentNo`, `parentAjaxId` are appended directly into the Location header without URL-encoding
  - Tomcat strips CR/LF so classic response-splitting is blocked, but parameter-reflection attacks downstream are possible
  - **Fix**: URL-encode each appended param via `Encode.forUriComponent()` — defensive hardening

- [ ] **D3 — `admin/manageFlowsheetsUpload.jsp:40`** (🟡 hardening)
  - Admin-tag uses `rights="r"` (read) for a state-changing upload operation
  - Semantically should be `rights="w"` (write) to match the destructive intent
  - Practical risk depends on RBAC grants (`_admin.flowsheet:r` rarely granted to untrusted accounts), but the code reads as more permissive than intended
  - **Fix**: change `rights="r"` to `rights="w"` in the admin-tag

- [ ] **D4 — `PathValidationUtils.validateExistingPath`** (🆕 adversarial, theoretical TOCTOU)
  - Method returns the **non-canonical** `File` object (line 106) while performing containment check against the canonical path
  - Callers (e.g. `FaxManagerImpl.resolveAndValidateFilePath`) then open via `file.toPath().normalize()` — lexical, not canonical
  - TOCTOU window between canonical validation and open
  - Practical risk: low (requires local filesystem write in the allowed directory tree)
  - **Fix**: return `file.getCanonicalFile()` instead of `file`; update the sibling `validateSource` method which already does this correctly (line 212/228/233)

---

### Category E — Systematic patterns requiring broader sweeps

- [ ] **E1 — Project-wide audit for `<%= expr.get\w+\(\) %>` without `Encode.*` wrapper**
  - The adversarial audit found ~10 additional XSS sinks beyond the ones listed above; these are the ones where ALL emissions on the page needed encoding (sampled files only)
  - A systematic grep + manual review is warranted
  - Consider CI lint rule: fail on raw `<%= ... %>` of method calls in JSP; require `Encode.` or `${e:...}` EL
  - Recommended tool: `semgrep` custom rule for JSP scriptlet emission

- [ ] **E2 — Project-wide audit for `escapeXml="false"` on user-controllable data**
  - JSTL `<c:out escapeXml="false"/>` is used legitimately in some places (rendering pre-encoded HTML) but appears incorrectly elsewhere
  - Audit each occurrence; convert incorrect ones to `"true"` or replace with `${e:forHtml(...)}` EL functions

- [ ] **E3 — Project-wide audit for `<c:url>` with EL-interpolated values in URL path/query**
  - The `<c:url value="...${param.x}..."/>` pattern does NOT URL-encode interpolated values
  - Correct usage: pass values via nested `<c:param>` which JSTL URL-encodes
  - Likely widespread; a single codemod could fix many instances

---

## Overall tracking

| Category | Items | Status |
|---|---|---|
| Stored XSS (JSP encoding) | 6 files in A + 5 in #1635 + 2 in #1634 | 7 tracked, 6 here |
| Reflected XSS | 2 files (B1, B2) | 2 here |
| Crypto / key mgmt | 3 items (C1, C2, C3) | C1 new, C2/C3 severity notes |
| Authz / hardening | 4 items (D1–D4) | 4 here |
| Systematic sweeps | 3 items (E1–E3) | 3 here |

**Total new items in this meta-ticket**: 15 individual fixes + 3 sweep initiatives.

---

## Recommended next steps

1. **Merge #1634 first** — quickest wins, highest-impact per line (7 fixes, ~45 min)
2. **Merge #1635 next** — encoding hardening (7 fixes, ~30 min)
3. **Schedule A1–A6** — multi-field XSS fixes; can be parallelized, one PR per file
4. **Design C1** — crypto key management needs architectural discussion before coding
5. **Plan E1–E3 sweeps** — create dedicated sweep issues (can leverage semgrep rules for detection)

---

## Why these aren't in the original scanner output

The adversarial methodology found these because:
- Scanner rules generally flag **source sites** (session write, `Pattern.compile`, etc.) but not the **consumer sites** that actually render into a sink. Cross-file trace was needed to find the real XSS.
- Scanners do not detect **duplicate unpatched files** (e.g. `provider/obarchecklistedit_99_12.jsp` was unflagged while its `decision/antenatal/` sibling was patched)
- Scanner rules for **key management** (hardcoded keys, no KDF, no rotation) are less mature than primitive-usage rules; the primitive calls in `EncryptionUtils` are correct
- JSTL `<c:url>` / `escapeXml="false"` anti-patterns are hard to model statically; real exploitation requires tracing EL interpolation semantics

This ticket serves as the institutional memory for these gaps so they don't regress when the codebase evolves.

---

🤖 Based on 4-pass adversarial validation (April 2026). Companion to #1634 and #1635.
